### PR TITLE
Fix lambda arg parsing

### PR DESCRIPTION
**Pull request by frankeld**
# Description

Previously, the grammar incorrectly parsed lambdas when the first arg was a function type. This was a general conflict between parenthesized expressions, lambda expressions, and anonymous expressions (note that lambda and anonymous expressions naturally overlap). This was solved by indirectly editing precedence through concatenating function type arg tokens.

This solution went through many iterations to find one that didn't cause other conflicts.

### Previous test output:
```
(script [0, 0] - [3, 0]
  (expression_statement [0, 0] - [2, 2]
    (parenthesized_expression [0, 0] - [2, 1]
      (lambda_expression [0, 1] - [2, 1]
        (parameters [0, 1] - [0, 26]
          (ERROR [0, 2] - [0, 19]
            (parameters [0, 10] - [0, 12])
            (type_specifier [0, 14] - [0, 18]))
          (parameter [0, 20] - [0, 25]
            name: (variable [0, 20] - [0, 25])))
        return_type: (type_specifier [0, 28] - [0, 32])
        body: (compound_statement [0, 37] - [2, 1])))))
```

# Testing
- [x] Added specific test for error
- [x] Corpus passes overall (no unit test regression)
- [x] Decreased error count in real-life code

### CODE REVIEW

File: grammar.js [hunk 1]
```diff
@@ -451,12 +451,9 @@ const rules = {
...
     seq(
       rep($._type_modifier),
-      '(',
-      'function',
-      '(',
+      '(function(',
       com.opt(opt($.inout_modifier), $._type, opt($.variadic_modifier), ','),
-      ')',
-      ':',
+      '):',
```
Comment (antoniodejesusochoasolano) [inline grammar.js:456]: Unfortunately, concatenating the function_type_specifier tokens this way prevents some valid Hack from being parsed. Having spaces in between all these tokens is valid Hack that the generated parser won't be able to parse.

E.g. Hack can parse this, but the updated grammar can't
```
(( function () : void) $test): void ==> {};
```

Comment (antoniodejesusochoasolano) [reply grammar.js:456]: No ideas come to mind for a solution right away, but can mess around with this later and see if I have any suggestions. Curious to hear what you tried.

Comment (frankeld) [reply grammar.js:456]: I think regex whitespace matching will work. Let me try and see what happens.

File: grammar.js [hunk 2]
```diff
@@ -448,12 +448,9 @@ const rules = {
...
     seq(
       rep($._type_modifier),
-      '(',
-      'function',
-      '(',
+      /\(\s*function\s*\(/,
       com.opt(opt($.inout_modifier), $._type, opt($.variadic_modifier), ','),
-      ')',
-      ':',
+      /\)\s*:/,
```
Comment (antoniodejesusochoasolano) [inline grammar.js:453]: Okay I have one more for you. You can actually use comments between the function type specifier tokens. But I'm thinking we can merge this PR and we can create an issue to solve this separately. We should support whatever Hack supports, but any more than spaces between these tokens seems unlikely in real code.

I would verify that we don't increase the errors in the ocaml-tree-sitter tests (since they test a lot more code so they might hit the edge cases).

Comment (frankeld) [reply grammar.js:453]: I think we can alleviate the impact of this by being less strict with the changes. For this fix to work, I think we only need to merge two of the rule items. Ex. the rule below works and allows comments except between function keyword and first parenthesis:
```
function_type_specifier: $ =>
  seq(
    rep($._type_modifier),
    '(',
    /function\s*\(/,
    com.opt(opt($.inout_modifier), $._type, opt($.variadic_modifier), ','),
    ')',
    ':',
    field('return_type', $._type),
    ')',
  ),
```
These now work (though the second breaks syntax highlighting in VSCode):
```
(  ( /*test*/ function ( ) : void ) $test ) : void ==> {
    
};

(  ( function (  )  /*test*/  : void ) $test ) : void ==> {
    
};
```
But this still fails (and also breaks syntax highlighting ):
```
(  ( function /*test*/ (  ) : void ) $test ) : void ==> {
    
};
```

I think especially with that small of an issue area, I would be good with merging after the described update and documenting the issue. Thoughts?
Note: we can choose any two string tokens to break comment support between (aka `(function` or `function(` or `:(`).

It would also be fixed if anonymous_function_expression was ever deleted.

Comment (frankeld) [reply grammar.js:453]: Also I run against real-world code for each PR (albeit internal repos) to always verify the error count goes down. This fix did resolve all the related errors and didn't add any as far as I can tell. I can follow up more about this process in the issues you opened, but the short of it is that they run against the repos here: https://github.com/returntocorp/ocaml-tree-sitter-languages/blob/main/lang/hack/projects.txt

Comment (antoniodejesusochoasolano) [reply grammar.js]: The compromise seems good to merge. We can fix the last case separately.
